algorithms/linfa-clustering/src/dbscan/algorithm.rs

```diff
@@ -79,32 +79,43 @@ impl<F: Float, D: Data<Elem = F>> Transformer<&ArrayBase<D, Ix2>, Array1<Option<
     for DbscanHyperParams<F>
 {
     fn transform(&self, observations: &ArrayBase<D, Ix2>) -> Array1<Option<usize>> {
-        let mut cluster_memberships = Array1::from_elem(observations.dim().0, None);
+        let mut cluster_memberships = Array1::from_elem(observations.nrows(), None);
         let mut current_cluster_id = 0;
-        for (i, obs) in observations.axis_iter(Axis(0)).enumerate() {
+        for i in 0..observations.nrows() {
             if cluster_memberships[i].is_some() {
                 continue;
             }
             let (neighbor_count, mut search_queue) =
-                find_neighbors(&obs, observations, self.tolerance(), &cluster_memberships);
+                find_neighbors(i, observations, self.tolerance(), &cluster_memberships);
             if neighbor_count < self.minimum_points() {
                 continue;
             }
+            // Tracks whether a value is in the search queue to prevent duplicates
+            let mut search_found = vec![false; observations.nrows()];
```

Comment:
doesn't this add a memory complexity of O(n^2) to the algorithm by allocating a bitmask not only for the neighborhood but all points. I'm not really familiar with the algorithm though

Reply:
The `search_found` bitmask is instantiated once for each observation, but only one instance exists at a time because it gets dropped at the end of each iteration. The size of `search_found` is equal to the number of observations/points, which is what n is equal to.

Reply:
yes sure but you still do N^2 allocations there, you could just re-use the same vec in every pass by moving it out of the loop